### PR TITLE
install source selection

### PR DESCRIPTION
**Pull request by mwhudson**
This is going to be hard to test until there is an image made with the livecd-rootfs change I just uploaded which won't happen for a day or two I guess but someone can read the code at least :)

### CODE REVIEW

**Review by dbungert — CHANGES_REQUESTED** [review]

File: subiquity/server/controllers/filesystem.py [hunk 1]
```diff
@@ -217,13 +213,11 @@ async def POST(self, config: list):
...
 
-    async def guided_GET(self, min_size: int = None, wait: bool = False) \
-            -> GuidedStorageResponse:
+    async def guided_GET(self, wait: bool = False) -> GuidedStorageResponse:
         probe_resp = await self._probe_response(wait, GuidedStorageResponse)
         if probe_resp is not None:
             return probe_resp
-        if not min_size:
-            min_size = DEFAULT_MIN_SIZE_GUIDED
+        min_size = 2*self.app.base_model.source.current.size + (1 << 30)
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/server/controllers/filesystem.py:220]: Ah you mean, the min size should be in the catalog? that might make sense yes...

File: subiquity/models/source.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        log.debug("loaded %d sources from %r", len(self.sources), fp.name)
+        if self.current is None:
+            self.current = self.sources[0]
+
+    def render(self):
+        path = os.path.join(self._dir, self.current.path)
+        scheme = self.current.type
+        return {
+            'sources': {
+                'ubuntu00': '{scheme}://{path}'.format(
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/models/source.py:86]: no not really. I've been programming Python too long to pick up these new-fangled things quickly :)

File: subiquity/ui/views/source.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        initial = {}
+
+        for default in True, False:
+            for source in sorted(sources, key=lambda s: s.id):
+                if source.default != default:
+                    continue
+                i = 2
+                while (2 << i) < source.size:
+                    i += 1
+                size = humanize_size(align_up(source.size, 2 << (i-3)))
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/ui/views/source.py:51]: Oh yes, this bit is all a bit hackish. I meant to pull it out into a shared and tested function before review...

I don't think we have to be _too_ worried about the catalog being wrong, I mean we're pretty dependent on the path in it being correct, but validating that size is strictly positive is probably a good idea.

File: subiquity/server/controllers/source.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            return
+        with open(path) as fp:
+            self.model.load_from_file(fp)
+
+    def interactive(self):
+        if len(self.model.sources) <= 1:
+            return False
+        return super().interactive()
+
+    async def GET(self) -> SourceSelectionAndSetting:
```
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/server/controllers/source.py:46]: the source endpoint is only interactive if there is more than one source in the catalog. So if you get curl to dump headers, you should see x-override-status: skip here.

**Comment by mwhudson** [comment]
FWIW you can play with this with the impish dailies now.

File: subiquity/ui/views/source.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/ui/views/source.py:51]: on second thought - does this cause more problems than it solves to show a number here?
The UI claims that the minimized will only take 512MiB, but a crude df on the result puts it at 6GiB
Non-minimized claims 1GiB/6.4GiB actual.
Can we proactively explain the difference?

[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/ui/views/source.py:51]: those numbers include 4GiB swap file, btw.
TIL /usr/lib/firmware is 700MiB+.

File: subiquity/server/controllers/filesystem.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/server/controllers/filesystem.py:224]: I've put a comment in the code about this. I'd like to move forward with this work if you trust me to not forget about coming back to this...

File: subiquity/ui/views/source.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/ui/views/source.py]: I've just dropped displaying a number here for now.

**Comment by mwhudson** [comment]
So I've just rebased this to simplify it a bit and also add some tests. Let me know what you think!

File: subiquity/server/controllers/filesystem.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/server/controllers/filesystem.py:224]: I trust you :)

File: subiquity/ui/views/source.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dbungert) [reply subiquity/ui/views/source.py]: Sorry, I really like having the number, I just was dreading the bug report claiming that it was lies.

**Review by dbungert — APPROVED** [review]

File: subiquity/ui/views/source.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (mwhudson) [reply subiquity/ui/views/source.py]: Yeah we should probably display something. For the next branch!

**Pull request merged**